src/NServiceBus.Core/ObjectBuilder/Common/CommonObjectBuilder.cs

```diff
@@ -3,101 +3,70 @@ namespace NServiceBus
     using System;
     using System.Collections.Generic;
     using System.Linq;
+    using System.Threading;
     using ObjectBuilder;
     using ObjectBuilder.Common;
 
-    class CommonObjectBuilder : IBuilder, IConfigureComponents
+    sealed class CommonObjectBuilder : IBuilder, IConfigureComponents
     {
-        public CommonObjectBuilder(IContainer container)
+        readonly AsyncLocal<IBuilder> currentBuilder;
+        readonly IBuilder previous;
+
+        public CommonObjectBuilder(IContainer container, AsyncLocal<IBuilder> scope = null)
         {
             this.container = container;
+            currentBuilder = scope ?? new AsyncLocal<IBuilder>();
+            previous = currentBuilder.Value;
+            currentBuilder.Value = this;
         }
 
-        public IBuilder CreateChildBuilder()
-        {
-            return new CommonObjectBuilder(container.BuildChildContainer());
-        }
+        // returns the current builder or the root
+        public IBuilder Current => currentBuilder.Value ?? this;
```

Comment:
can the value ever be null? (except when the root container is being disposed)

Reply:
Yes. I mentioned this yesterday and that was also one of the things you seem to have been observing before is during startup the scope is not entirely deterministic that's why it is the best to fall back to the root which is this when there is no child. 

See all the tests that failed with NullRef

https://github.com/Particular/NServiceBus/actions/runs/2695891787